handler/pullRequestHandler.go

```diff
@@ -45,56 +46,60 @@ func HandlePullRequest(req types.PullRequestOuter, contributingURL string, confi
-
-		if labelErr != nil {
-			log.Fatalln(labelErr)
-		}
-		fmt.Println("Current labels ", issue.Labels)
-
-		if hasNoDcoLabel(issue) == false {
-			fmt.Println("Applying label")
-			_, res, assignLabelErr := client.Issues.AddLabelsToIssue(ctx, req.Repository.Owner.Login, req.Repository.Name, req.PullRequest.Number, []string{"no-dco"})
-			if assignLabelErr != nil {
-				log.Fatalf("%s limit: %d, remaining: %d", assignLabelErr, res.Limit, res.Remaining)
-			}
+		log.Fatalf("unable to fetch pull request commits for PR %d: %s", req.PullRequest.Number, err)
+	}
 
-			body :=
-				`Thank you for your contribution. I've just checked and your commit doesn't appear to be signed-off. That's something we need before your Pull Request can be merged. Please see our [contributing guide](` + contributingURL + `).
-Tip: if you only have one commit so far then run: ` + "`" + `git commit --amend --sign-off` + "`" + ` and then ` + "`" + `git push --force` + "`."
+	issue, res, labelErr := client.Issues.Get(ctx, req.Repository.Owner.Login, req.Repository.Name, req.PullRequest.Number)
+	if labelErr != nil {
+		log.Fatalf("unable to fetch labels for PR %d: %s", req.PullRequest.Number, err)
+	}
+	fmt.Println("Rate limiting", res.Rate)
+	res.Body.Close()
 
-			comment := &github.IssueComment{
-				Body: &body,
-			}
+	anonymousSign := hasAnonymousSign(commits)
+	unsignedCommits := hasUnsigned(commits)
+	noDcoLabelExists := hasNoDcoLabel(issue)
 
-			comment, resp, err := client.Issues.CreateComment(ctx, req.Repository.Owner.Login, req.Repository.Name, req.PullRequest.Number, comment)
-			if err != nil {
-				log.Fatalf("%s limit: %d, remaining: %d", assignLabelErr, resp.Limit, resp.Remaining)
-				log.Fatal(err)
+	if !anonymousSign && !unsignedCommits {
+		fmt.Println("Things look OK right now.")
+		if noDcoLabelExists {
+			fmt.Println("Removing label")
+			res, removeLabelErr := client.Issues.RemoveLabelForIssue(ctx, req.Repository.Owner.Login, req.Repository.Name, req.PullRequest.Number, "no-dco")
+			if removeLabelErr != nil {
+				log.Fatalf("unable to remove DCO label from PR %d: %s", req.PullRequest.Number, err)
 			}
-			fmt.Println(comment, resp.Rate)
+			fmt.Println("Rate limiting", res.Rate)
+			res.Body.Close()
 		}
+		return
+	}
+
+	var body string
+	if anonymousSign {
+		body =
+			`Thank you for your contribution. It seems that one or more of your commits have an anonymous email address. Please consider signing your commits with a valid email address. Please see our [contributing guide](` + contributingURL + `).`
 	} else {
-		fmt.Println("Things look OK right now.")
-		issue, res, labelErr := client.Issues.Get(ctx, req.Repository.Owner.Login, req.Repository.Name, req.PullRequest.Number)
+		body =
+			`Thank you for your contribution. I've just checked and your commit doesn't appear to be signed-off. That's something we need before your Pull Request can be merged. Please see our [contributing guide](` + contributingURL + `).
```

Comment:
For this and the other message please extract a function to generate the message. Nothing fancy, one function per message.

Reply:
It will help with readability and we can also test it.

Reply:
No problem :+1: